homeassistant/components/opower/coordinator.py

```diff
@@ -0,0 +1,205 @@
+"""Coordinator to handle Opower connections."""
+from datetime import datetime, timedelta
+import logging
+from types import MappingProxyType
+from typing import Any, cast
+
+from aiohttp.client_exceptions import ClientResponseError
+from opower import Account, AggregateType, CostRead, Forecast, MeterType, Opower
+
+from homeassistant.components.recorder import get_instance
+from homeassistant.components.recorder.models import StatisticData, StatisticMetaData
+from homeassistant.components.recorder.statistics import (
+    async_add_external_statistics,
+    get_last_statistics,
+    statistics_during_period,
+)
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME, UnitOfEnergy, UnitOfVolume
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import CONF_UTILITY, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class OpowerCoordinator(DataUpdateCoordinator):
+    """Handle fetching Opower data, updating sensors and inserting statistics."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry_data: MappingProxyType[str, Any],
+    ) -> None:
+        """Initialize the data handler."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Opower",
+            update_interval=timedelta(hours=12),
+        )
+        self.api = Opower(
+            aiohttp_client.async_get_clientsession(hass),
+            entry_data[CONF_UTILITY],
+            entry_data[CONF_USERNAME],
+            entry_data[CONF_PASSWORD],
+        )
+
+    async def _async_update_data(
+        self,
+    ) -> dict[str, Forecast]:
+        """Fetch data from API endpoint."""
+        try:
+            await self.api.async_login()
+        except ClientResponseError as err:
+            if err.status in (401, 403):
+                # Cancel future updates and start config flow config async_step_reauth
+                raise ConfigEntryAuthFailed from err
+            # Let DataUpdateCoordinator handle ClientError retries
+            raise err
+        forecasts = await self.api.async_get_forecast()
+        _LOGGER.debug("Updating sensor data with: %s", forecasts)
+        await self._insert_statistics([forecast.account for forecast in forecasts])
+        return {forecast.account.utility_account_id: forecast for forecast in forecasts}
+
+    async def _insert_statistics(self, accounts: list[Account]) -> None:
+        """Insert Opower statistics."""
+        if "recorder" not in self.hass.config.components:
```

Comment:
I don't think this is needed? You have recorder in dependencies in the manifest

Reply:
Removed. You are likely right. Tibber integration doesn't check but kitchen_sink does https://github.com/home-assistant/core/blob/dev/homeassistant/components/kitchen_sink/__init__.py#L52